### ISSUE TITLE
New AI system

### ISSUE THREAD

**Issue by exectails** [aura-project/aura#171]
I'm planning to add a new AI system in the future, that works a little bit different from the current one. The main goal will be to make it easier to maintain, and to clean up the entire system.

More information: http://aura-project.org/forum/topic/124-ai-and-npc-client/

This shouldn't discourage anyone from adding currently missing AIs, as writing said AIs gives us a lot of information on what's wrong about the current system, and what needs to be fixed or added to the new one. (If you notice anything, please mention it here or on the forum.)

Once the new system is done, I will convert the existing AIs.

**Currently missing AI features**
- Stackable skills
- Hearing (do we want/need this in Aura?)
- Aggro nearby over time
- Aggro on attacking friendlies
- Weapon swapping

**Current bugs/glitches**
- When attacking multiple times, the creature will keep using the same skill, even though it should be canceled after the first time.
- The AI sometimes seems to be very fast with loading skills and attacking, as if they didn't have a stun.

**Comment by Sairii** [comment]
I think hearing is important. But we could always add this later?
Worm as example has very poor visual aggro 600 in 45° and 400 audio.
If only this game had backstab...

**Comment by exectails** [comment]
Once you know hearing exists, you start to think Aura should have it as well to function official-like, but I find it funny how that feature was seemingly unknown until recently. I've never heard anyone mention this, nobody on the chat or forum seemed to know anything about it, you had to test it to confirm its existence. Nobody would ever know =X

But yea, I guess we should add it...

**Comment by Sairii** [comment]
I always noticed it but was never really 100% sure.
Sometimes a wolf just aggroed from miles away even tho I thought its aggro range was smaller.

Well it really depends on how hard it would be to implement.
So its your decision.

**Comment by exectails** [comment]
It's certainly easier to not include it, but it would forever bug us, now that we know for sure =P

**Comment by ImaBrokeDude** [comment]
Remember when the AI attack speed and range would shrink and grow exponentially? We have come a long way :D

**Comment by Sairii** [comment]
>The AI sometimes seems to be very fast with loading skills and attacking, as if they didn't have a stun.

There was also a problem with an enemy doing Attack combos and the player being knocked down.
Once the player is knocked down the AI should stop pursuing its Attack combo.

On Aura the enemy would wait for the player to stand back up and then finish its melee combo (like Do Attack 3, -- 2 attacks and player is down, waits till player is back on feet, does the last 1 attack)
Official it stops on knockdown. Waits (?? not sure how long) once the player is down and then goes back to the onaggro AI parts.

**Comment by Pegwald** [comment]
> On Aura the enemy would wait for the player to stand back up and then finish its melee combo (like Do Attack 3, -- 2 attacks and player is down, waits till player is back on feet, does the last 1 attack)

Yeah, if the enemy wants to do a 3-hit combo, you can defend the first one and take two open hits without being knocked down. This way you can do things like defend>charge magic>get hit twice>cast magic. I Think the magic is only canceled if you're knocked down. Is that how it's supposed to be or is that a bug too? ^^''
